### ISSUE TITLE
Document batch_size argument to Monitor.setup

### ISSUE THREAD

**Issue by dwf** [lisa-lab/pylearn2#874]

**Comment by lamblin** [comment]
We should also clarify the relationship between batch_size and force_batch size, in Monitor as well as in SGD and Model.
- [ ] dig out the thread discussing that in pylearn-dev
- [ ] decide how things should work and make a plan
- [ ] implement it

**Comment by bouthilx** [comment]
batch_size and num_batches are only used for Monitor.add_dataset() in Monitor.setup(). batch_size should be set to None by default so that we can pass either batch_size or num_batches. 

Right now, force_batch_size is automatically set to batch_size in MLP. When batch_size is set in MLP, it forces SGD to use MLP.batch_size unless SGD is able to override MLP.batch_size (happens in TrainingAlgorithm._synchronize_batch_size()). 

batch_size in models is needed for those who need a fix batch size. Thus, training algorithms should be forced to use their batch size. It seems to me that the interface of models should only have force_batch_size and TrainingAlgorithms is the only responsible for batch_size. Another options would be to keep batch_size and change force_batch_size as a boolean argument and disable set_batch_size when force_batch_size is True. If however force_batch_size is false, batch_size in TrainingAlgorithms takes precedence. 

@lamblin What do you think about that based on what you have seen while fixing maxout paper yaml files?

**Comment by lamblin** [comment]
For reference, discussion started at https://groups.google.com/d/topic/pylearn-dev/NQyq7a6AExg/discussion

I would agree with your first solution for Models that are trained by SGD, but the DefaultTrainingAlgorithm, for instance, depends on a batch_size set by the Model, since in that case the Model knows how to train itself.
We could deprecate `batch_size` in favor of `force_batch_size` at least for MLP and Layers. We should also go through the examples and tutorials to favor specifying the batch size through SGD.

**Comment by bouthilx** [comment]
Where does it rely on model batch size? What I see is model.model.train_batch(dataset, batch_size) at line 141. We should try to keep interface uniform for all models and training algorithms if possible.

**Comment by lamblin** [comment]
@bouthilx You're right, let's deprecate batch_size, replace it by force_batch_size, and remove it from the examples.

**Comment by bouthilx** [comment]
@lamblin SGTM. Before, we should look at the scripts to see if they are using batch size. If I remember correctly there was some script to produce predictions for Kaggle that was using model.batch_size.

**Comment by lamblin** [comment]
Also, apparently, `DBM` makes use of `self.batch_size`, that is considered constant, and does not work without a specified batch size at the model level.
